File: problems/011-mount-tree-conflicts/solutions/python/main.py
import sys
from array import array


tokens = sys.stdin.buffer.read().split()
n = int(tokens[0])
infinity = n + 1
first_child = array("i", [0])
next_sibling = array("i", [0])
exact_min = array("i", [infinity])
file_min = array("i", [infinity])
desc_min = array("i", [infinity])
node_char = bytearray(1)


def find_or_create_child(parent, char):
    child = first_child[parent]
    while child:
        if node_char[child] == char:
            return child
        child = next_sibling[child]
    child = len(node_char)
    first_child.append(0)
    next_sibling.append(first_child[parent])
    exact_min.append(infinity)
    file_min.append(infinity)
    desc_min.append(infinity)
    node_char.append(char)
    first_child[parent] = child
    return child


token_at = 1
for j in range(1, n + 1):
    kind = tokens[token_at]
    path = tokens[token_at + 1]
    token_at += 2
    visited = []
    current = 0
    best = infinity

    for position, char in enumerate(path):
        current = find_or_create_child(current, char)
        visited.append(current)
        if position + 1 < len(path) and (
            position == 0 or path[position + 1] == ord("/")
        ):
            best = min(best, file_min[current])
    best = min(best, exact_min[current])
    if kind == b"F":
        best = min(best, desc_min[current])

    if best != infinity:
        print("CONFLICT", best, j)
        raise SystemExit

    for position in range(len(path) - 1):
        if position == 0 or path[position + 1] == ord("/"):
            at = visited[position]
            desc_min[at] = min(desc_min[at], j)
    exact_min[current] = min(exact_min[current], j)
    if kind == b"F":
        file_min[current] = min(file_min[current], j)

print("VALID")

File: problems/011-mount-tree-conflicts/solutions/rust/main.rs
use std::io::{self, Read};

#[derive(Clone, Copy)]
struct Node {
    first_child: u32,
    next_sibling: u32,
    exact_min: u32,
    file_min: u32,
    desc_min: u32,
    ch: u8,
}

fn find_or_create_child(nodes: &mut Vec<Node>, parent: u32, ch: u8, infinity: u32) -> u32 {
    let mut child = nodes[parent as usize].first_child;
    while child != 0 {
        if nodes[child as usize].ch == ch {
            return child;
        }
        child = nodes[child as usize].next_sibling;
    }
    let child = nodes.len() as u32;
    let next_sibling = nodes[parent as usize].first_child;
    nodes.push(Node {
        first_child: 0,
        next_sibling,
        exact_min: infinity,
        file_min: infinity,
        desc_min: infinity,
        ch,
    });
    nodes[parent as usize].first_child = child;
    child
}

fn main() {
    let mut input = Vec::new();
    io::stdin().read_to_end(&mut input).unwrap();
    let text = std::str::from_utf8(&input).unwrap();
    let mut tokens = text.split_ascii_whitespace();
    let n: u32 = tokens.next().unwrap().parse().unwrap();
    let infinity = n + 1;
    let mut nodes = vec![Node {
        first_child: 0,
        next_sibling: 0,
        exact_min: infinity,
        file_min: infinity,
        desc_min: infinity,
        ch: 0,
    }];

    for j in 1..=n {
        let kind = tokens.next().unwrap().as_bytes()[0];
        let path = tokens.next().unwrap().as_bytes();
        let mut visited = Vec::with_capacity(path.len());
        let mut current = 0_u32;
        let mut best = infinity;

        for (position, &ch) in path.iter().enumerate() {
            current = find_or_create_child(&mut nodes, current, ch, infinity);
            visited.push(current);
            if position + 1 < path.len() && (position == 0 || path[position + 1] == b'/') {
                best = best.min(nodes[current as usize].file_min);
            }
        }
        best = best.min(nodes[current as usize].exact_min);
        if kind == b'F' {
            best = best.min(nodes[current as usize].desc_min);
        }

        if best != infinity {
            println!("CONFLICT {best} {j}");
            return;
        }

        for position in 0..path.len() - 1 {
            if position == 0 || path[position + 1] == b'/' {
                let node = visited[position] as usize;
                nodes[node].desc_min = nodes[node].desc_min.min(j);
            }
        }
        nodes[current as usize].exact_min = nodes[current as usize].exact_min.min(j);
        if kind == b'F' {
            nodes[current as usize].file_min = nodes[current as usize].file_min.min(j);
        }
    }

    println!("VALID");
}
